server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java

```diff
@@ -133,11 +134,34 @@ public static DocIdAndVersion loadDocIdAndVersion(IndexReader reader, Term term,
         return null;
     }
 
+    public static DocIdAndVersion loadDocIdAndVersion(IndexReader reader, Term term, long timestamp, boolean loadSeqNo) throws IOException {
+        PerThreadIDVersionAndSeqNoLookup[] lookups = getLookupState(reader, term.field(), true);
+        List<LeafReaderContext> leaves = reader.leaves();
+        // iterate in default order, the segments should be sorted by DataStream#TIMESERIES_LEAF_READERS_SORTER
```

Comment:
Could tsdb be applied to non-datastream indices (even if not intended)? If so, we need to convert some of the assertions to `if` statements instead.

Reply:
I think yes. The requirement is that an Index has an `@timestamp` field and the leaf readers are sorted by the max timestamp. This is also true in case of regular data streams. The `TIMESERIES_LEAF_READERS_SORTER` is used when `mapperService.mappingLookup().hasTimestampField()` returns `true`.

Reply:
There is another detail here that is important. The timestamp is parsed from the id and currently only in the case of tsdb does the id contain the timestamp. This isn't the case for regular data streams. 

I think for now we need the index mode check.

Reply:
Sure. I think of this assertion:

```
assert prevMaxTimestamp >= lookup.maxTimestamp;
```

which will only be valid for data streams, not for tsdb used against regular indices (I think).

Reply:
> which will only be valid for data streams, not for tsdb used against regular indices (I think).

I think this will be valid when `DataStream#TIMESERIES_LEAF_READERS_SORTER` is used and
currently this is used for all indices that have a `@timestamp` field (see `MappingLookup#hasTimestampField()`).
So it isn't tight to data streams.

Reply:
Sorry for being slow, looks like you are right. The level of intimacy across many components is tricky here, but that is mostly unrelated to your PR.